.pre-commit-config.yaml

```diff
@@ -1,11 +1,69 @@
 repos:
--   repo: https://github.com/psf/black
-    rev: 22.10.0
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.5.0
     hooks:
-    -   id: black
-        exclude: 'docs|examples|case-studies|tests'
--   repo: https://github.com/pycqa/flake8
-    rev: 6.0.0
+      - id: trailing-whitespace
+      - id: end-of-file-fixer
+      - id: check-json
+      - id: pretty-format-json
+      - id: debug-statements
+
+  - repo: https://github.com/asottile/setup-cfg-fmt
+    rev: v2.4.0
+    hooks:
+      - id: setup-cfg-fmt
+
+  - repo: https://github.com/MarcoGorelli/absolufy-imports
+    rev: v0.3.1
+    hooks:
+      - id: absolufy-imports
+
+  - repo: https://github.com/pycqa/isort
+    rev: 5.12.0
+    hooks:
+      - id: isort
+        name: isort (python)
+        args: ["--profile", "black", "-l", "79"]
+
+  - repo: https://github.com/asottile/pyupgrade
+    rev: v3.10.1
     hooks:
-    -   id: flake8
+      - id: pyupgrade
+        args: [--py39-plus]
```

Comment:
This will automatically switch things to 3.9 syntax, right?  I'm supportive of this, but just wanted to point it out as part of my justification for removing 3.8 from the test matrix.

Reply:
I actually left this in accidentally. It's not crucial but I guess aligns practices between CBI (>=3.12) and P3 (3.9, 3.10); I can't figure out from reading [pyupgrade docs](https://github.com/asottile/pyupgrade?tab=readme-ov-file#subprocessrun-replace-universal_newlines-with-text) whether every version we bump is inclusive (i.e. 3.9 includes 3.7, 3.8, etc.). I would assume so, but nominally we would want to configure this in a way that matches the Python version we're using.